buildkite/src/Lib/SelectFiles.dhall

```diff
@@ -0,0 +1,86 @@
+-- A DSL for representing file globbing; compiles into egrep-compatible regexp
+let Prelude = ../External/Prelude.dhall
+let P = Prelude
+
+let PathPattern = < Lit : Text | Any > -- Any == *
+let FilePattern = {
+  Type = { dir: Optional (List PathPattern), exts: Optional (List Text) },
+  default = { dir = None (List PathPattern), exts = None (List Text) }
+}
+
+let everything : FilePattern.Type = FilePattern.default
+
+let contains : Text -> FilePattern.Type =
+  \(s : Text) ->
+  { dir = Some [PathPattern.Lit s],
+    exts = None (List Text)
+  }
+
+let exactly : Text -> Text -> FilePattern.Type =
+  \(dir : Text) -> \(ext : Text) ->
+    { dir = Some [PathPattern.Lit dir],
+      exts = Some [ext]
+    }
+
+-- compile things
+let any : List Text -> Text =
+  P.Text.concatSep "|"
+
+let compileExt : Text -> Text =
+  \(ext : Text) -> "\\." ++ ext ++ "\\\$"
+
+let compileExts : List Text -> Text =
+  \(exts : List Text) ->
+    let exts_ = P.List.map Text Text compileExt exts in
+    "(" ++ (any exts_) ++ ")"
+
+let compilePath : PathPattern -> Text =
+  \(path : PathPattern) ->
+    merge {
+      Lit = \(text : Text) -> text,
+      Any = ".*"
+    } path
+
+let compilePaths : List PathPattern -> Text =
+  \(paths : List PathPattern) ->
+    let paths_ = P.List.map PathPattern Text compilePath paths in
+    P.Text.concat paths_
+
+let compileFile : FilePattern.Type -> Text =
+  \(pattern : FilePattern.Type) ->
+    let dirPart =
+      P.Optional.fold (List PathPattern) pattern.dir Text compilePaths ".*"
+    in
+    let extPart =
+      P.Optional.fold (List Text) pattern.exts Text compileExts ""
+    in
+    dirPart ++ extPart
+
+let compile : List FilePattern.Type -> Text =
+  \(patterns : List FilePattern.Type) ->
+    let patterns_ = P.List.map FilePattern.Type Text compileFile patterns in
```

Comment:
Part of the compilation should also involve escaping invalid characters in the literals (`.`, `*`, `^`, `$`, etc). Escaping `.` here is certainly important, and it's good to just add escapes for all (or most) of the special regexp characters to help future proof this.

Reply:
See above. I wish we could, but I think it will be a benign error if we can't.